### PR TITLE
feat: Multi-GPU Data Parallel (DDP) for training

### PR DESCRIPTION
**Pull request by aniruddhaadak80**
### Summary
Introduces `torch.distributed` (DDP) setup to natively run the training script on multi-GPU setups using `torchrun`. 

### Key Capabilities
- **Effortless Multi-Device Sync:** Adapts the existing single-file `train.py` architecture into a data-parallel approach.
- **Dynamic Adjustments:** Automatically adjusts batch accumulations and assigns `device=cuda:local_rank` to map perfectly to underlying hardware topology.
- **DDP Micro-Step Optimization:** Limits `require_backward_grad_sync` strictly to the terminal micro-steps, massively increasing throughput when using gradient accumulation to not stall compute networking.
- **Rank 0 Isolation:** Suppresses output logs on non-master devices (`builtins.print` disabled) to prevent terminal flooding.

Runs perfectly out of the box with zero modifications required inside `prepare.py`.

### Usage
```bash
torchrun --nproc_per_node=4 train.py
```

### CODE REVIEW

**Comment by aniruddhaadak80** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/karpathy/autoresearch/pull/117#pullrequestreview-3920449975)
